std/internal/math/biguintcore.d

```diff
@@ -345,35 +346,45 @@ public:
     }
 
     // return false if invalid character found
-    bool fromHexString(const(char)[] s) pure nothrow @safe
+    bool fromHexString(Range)(Range s) if (
+        isBidirectionalRange!Range && isSomeChar!(ElementType!Range))
     {
+        import std.range : walkLength;
+
         //Strip leading zeros
-        int firstNonZero = 0;
-        while ((firstNonZero < s.length - 1) &&
-            (s[firstNonZero]=='0' || s[firstNonZero]=='_'))
-        {
-                ++firstNonZero;
-        }
-        auto len = (s.length - firstNonZero + 15)/4;
-        auto tmp = new BigDigit[len+1];
-        uint part = 0;
-        uint sofar = 0;
-        uint partcount = 0;
-        assert(s.length>0);
-        for (ptrdiff_t i = s.length - 1; i>=firstNonZero; --i)
-        {
-            assert(i>=0);
-            char c = s[i];
-            if (s[i]=='_') continue;
-            uint x = (c>='0' && c<='9') ? c - '0'
-                   : (c>='A' && c<='F') ? c - 'A' + 10
-                   : (c>='a' && c<='f') ? c - 'a' + 10
+        while (!s.empty && s.front == '0')
+        {
+            s.popFront;
+        }
+
+        if (s.empty)
+        {
+            data = ZERO;
+            return true;
+        }
+
+        auto len = (s.save.walkLength + 15) / 4;
+        auto tmp = new BigDigit[len + 1];
+        uint part, sofar, partcount;
+
+        foreach_reverse (character; s)
+        {
+            if (character == '_')
+                continue;
+
+            uint x = (character>='0' && character<='9') ? character - '0'
+                   : (character>='A' && character<='F') ? character - 'A' + 10
+                   : (character>='a' && character<='f') ? character - 'a' + 10
                    : 100;
-            if (x==100) return false;
+
+            if (x == 100)
+                return false;
```

Comment:
This use of `100` as a nonsense value seems to be needless. Why not just break down the long-winded series of `?:`'s into a series of if-else statements ('cos that's what they get translated into anyway)? It will be easier to read and doesn't leave people wondering where on earth the `100` came from.